ActiveShardsObserverIT#testCreateIndexNotEnoughActiveShardsTimesOut fails

The test failed only once in CI:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+java9-periodic/673/

The test itself succeeds actually, the problem is in the after test checks:

```
FAILURE 5.30s J1 | ActiveShardsObserverIT.testCreateIndexNotEnoughActiveShardsTimesOut <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: Shard [test-idx][6] is still locked after 5 sec waiting
   > 	at __randomizedtesting.SeedInfo.seed([CAB167E7B91DDDA2:F4E0B8D9B6F84C47]:0)
   > 	at org.elasticsearch.test.InternalTestCluster.assertAfterTest(InternalTestCluster.java:1913)
   > 	at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:550)
   > 	at org.elasticsearch.test.ESIntegTestCase.after(ESIntegTestCase.java:1980)
   > 	at jdk.internal.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
   > 	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-ea/DelegatingMethodAccessorImpl.java:43)
   > 	at java.lang.Thread.run(java.base@9-ea/Thread.java:843)
```

I failed to repro this locally with same seed and many iterations.

--- a/core/src/test/java/org/elasticsearch/action/support/ActiveShardsObserverIT.java
+++ b/core/src/test/java/org/elasticsearch/action/support/ActiveShardsObserverIT.java
@@ -21,6 +21,7 @@ package org.elasticsearch.action.support;
 
 import org.elasticsearch.action.ListenableActionFuture;
 import org.elasticsearch.action.admin.indices.create.CreateIndexResponse;
+import org.elasticsearch.common.Priority;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.test.ESIntegTestCase;
 
@@ -44,12 +45,14 @@ public class ActiveShardsObserverIT extends ESIntegTestCase {
             settingsBuilder.put("index.routing.allocation.exclude._name", exclude);
         }
         Settings settings = settingsBuilder.build();
-        assertFalse(prepareCreate("test-idx")
+        final String indexName = "test-idx";
+        assertFalse(prepareCreate(indexName)
                        .setSettings(settings)
                        .setWaitForActiveShards(randomBoolean() ? ActiveShardCount.from(1) : ActiveShardCount.ALL)
                        .setTimeout("100ms")
                        .get()
                        .isShardsAcked());
+        waitForIndexCreationToComplete(indexName);
     }
 
     public void testCreateIndexNoActiveShardsNoWaiting() throws Exception {
@@ -74,22 +77,24 @@ public class ActiveShardsObserverIT extends ESIntegTestCase {
         final int numReplicas = numDataNodes + randomInt(4);
         Settings settings = Settings.builder()
                                 .put(indexSettings())
-                                .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(1, 7))
+                                .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(1, 5))
                                 .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), numReplicas)
                                 .build();
-        assertFalse(prepareCreate("test-idx")
+        final String indexName = "test-idx";
+        assertFalse(prepareCreate(indexName)
                        .setSettings(settings)
                        .setWaitForActiveShards(ActiveShardCount.from(randomIntBetween(numDataNodes + 1, numReplicas + 1)))
                        .setTimeout("100ms")
                        .get()
                        .isShardsAcked());
+        waitForIndexCreationToComplete(indexName);
     }
 
     public void testCreateIndexEnoughActiveShards() throws Exception {
         final String indexName = "test-idx";
         Settings settings = Settings.builder()
                                 .put(indexSettings())
-                                .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(1, 7))
+                                .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(1, 5))
                                 .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), internalCluster().numDataNodes() + randomIntBetween(0, 3))
                                 .build();
         ActiveShardCount waitForActiveShards = ActiveShardCount.from(randomIntBetween(0, internalCluster().numDataNodes()));
@@ -104,12 +109,17 @@ public class ActiveShardsObserverIT extends ESIntegTestCase {
                                 .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(1, 5))
                                 .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), numReplicas)
                                 .build();
-        assertFalse(prepareCreate("test-idx1")
+        final String indexName = "test-idx";
+        assertFalse(prepareCreate(indexName)
                        .setSettings(settings)
                        .setWaitForActiveShards(ActiveShardCount.ALL)
                        .setTimeout("100ms")
                        .get()
                        .isShardsAcked());
+        waitForIndexCreationToComplete(indexName);
+        if (client().admin().indices().prepareExists(indexName).get().isExists()) {
+            client().admin().indices().prepareDelete(indexName).get();
+        }
 
         // enough data nodes, all shards are active
         settings = Settings.builder()
@@ -117,7 +127,7 @@ public class ActiveShardsObserverIT extends ESIntegTestCase {
                         .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(1, 7))
                         .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), internalCluster().numDataNodes() - 1)
                         .build();
-        assertAcked(prepareCreate("test-idx2").setSettings(settings).setWaitForActiveShards(ActiveShardCount.ALL).get());
+        assertAcked(prepareCreate(indexName).setSettings(settings).setWaitForActiveShards(ActiveShardCount.ALL).get());
     }
 
     public void testCreateIndexStopsWaitingWhenIndexDeleted() throws Exception {
@@ -143,6 +153,14 @@ public class ActiveShardsObserverIT extends ESIntegTestCase {
 
         logger.info("--> ensure the create index request completes");
         assertAcked(responseListener.get());
+    }
+
+    // Its possible that the cluster state update task that includes the create index hasn't processed before we timeout,
+    // and subsequently the test cleanup process does not delete the index in question because it does not see it, and
+    // only after the test cleanup does the index creation manifest in the cluster state.  To take care of this problem
+    // and its potential ramifications, we wait here for the index creation cluster state update task to finish
+    private void waitForIndexCreationToComplete(final String indexName) {
+        client().admin().cluster().prepareHealth(indexName).setWaitForEvents(Priority.URGENT).get();
     }
 
 }

